### DISCUSSION TITLE
Send some information when Metals "stuck"s

### DISCUSSION THREAD

**Discussion by tanishiking** [scalameta/metals#4476] [General]
https://discord.com/channels/632642981228314653/632652693013528589/1022083593125756938

> velvetbaldmime — 09/21/2022
I have a file at work, which is in an interesting state - the compilation is broken, and the semanticdb file doesn't exist.
>
> What this means is that there's no completions, and no information as to why there are no completions. Just no reaction to anything
>
> For someone who doesn't trace LSP logs it might look like Metals just died - but then it recompiles on save it published diagnostics.
> The project is tiny, but Bloop takes 90s and seems to compile it from scratch all the time - so you can imagine how annoying it is trying to fix compilation issues having no help from the IDE.
> The questions are:
> 1. Why do the semanticdb get deleted so aggressively?
> 2. Is there any way to get completions (any without semanticdb files)?
> 3. Can Metals indicate as to why is cannot offer any completions?
> Because I presume that at the time of "complete" request Metals is aware why it might not be able to offer anything.
>
> The project is Scala 2.13, build is setup with Maven, so BSP is Bloop.
> I also checked - once the compilation finishes, the semanticdb for that file is deleted again if there are any errors - so with a feedback loop lasting a whole minute you basically never have any completions.
> > 3. Can Metals indicate as to why is cannot offer any completions?
> perhaps via file-level warning diagnostic?

> Tomasz Godzik — 09/21/2022
> 1. We don't actually remove the files even if the compilation failed.
> 2. Completions only use the presentation compiler, but needs classpath to work with things outside of the file.
> 3. There should actually be some exceptions in metals.log, but this should be super rare

> velvetbaldmime — 09/21/2022
> (1) I actually checked that - the semanticdb for that file appears and then disappears immediately
> The "show semanticdb" command says as much - that it doesn't exist.

---

https://discord.com/channels/632642981228314653/632652693013528589/1026553712778412052

> mhughes — Today at 2:57 AM
> What diagnostics could be added to Metals to help in the scenario where Intellisense just stops working?  I've developed a pretty good intuition for when a restart of Metals is enough vs when I need to just nuke everything (.metals/.bloop) but I really wish we had some way to knowing "what is going on inside VSCode extension / Metals / Bloop (for compile issues)".  Obviously the ultimate goal is for everything to "just work" but between me, my teammates, and some vocal folks on Twitter, sometimes "Metals" (loosely using that to cover VsCode/Metals/build server) stops working and I would love to help diagnose these issues when they appear but beyond scoping out the bsp/lsp trace files I don't know what else to look for.

> velvetbaldmime — Today at 3:42 AM
> I had the same question a while back: https://discord.com/channels/632642981228314653/632652693013528589/1022083784050495498
> 
> The issue was of similar nature - absence of semanticdb leads to 0 information from Metals.
> There isn't much room in the LSP protocol for global information - but IMO metals should sent some information back via window/showMessage at least

related https://github.com/scalameta/metals/discussions/3769

**Comment by tgodzik** [comment]
I think we should enable these information under the debug flag. I am also thinking that we could automatically create the trace files is the debug flag is on :thinking: 

Also, we should log some more information from the compiler to see why it's missing completions

**Reply by tgodzik** [reply]
Though coming to think of it most issues show up in `.metals/metals.log`, I wonder what is happening with the mentioned project that so many things go so badly. It might be an issue with incremental compilation that over aggressively invalidates files and in that case Presentation Compiler will not be able to provide completions since classfiles are removed :thinking: 

Though, we might want to add some more debug data for all the cases that Metals failed to provide results not only completions. Will try to add things next time we change any relevant part.
